c/tests/check_grounding_space.c

```diff
@@ -97,12 +103,291 @@ START_TEST (test_replace)
 }
 END_TEST
 
+START_TEST (test_abstract_space_grounding)
+{
+    space_t* space = space_new_grounding_space();
+    space_add(space, expr(atom_sym("+"), atom_var("a"), atom_sym("B"), 0));
+    atom_t* query = expr(atom_sym("+"), atom_sym("A"), atom_var("b"), 0);
+
+    struct output_t result = { "", 0 };
+    space_query(space, query, query_callback, &result);
+    ck_assert_str_eq(result.str, "b: B, ");
+
+    space_free(space);
+}
+END_TEST
+
+void str_to_buf(const char *str, void *context) {
+    strncpy(context, str, BUF_SIZE);
+};
+
+// QUESTION FOR VITALY: What's the simplest way to test the observer mechanism?  Obviously I could
+// add my own observer, but I assume there is a better in-situ test that is possible?
+typedef struct _observer_list_item {
+    space_observer_t* observer;
+    struct _observer_list_item* next;
+} observer_list_item;
+
+void notify_all_observers(observer_list_item* list, space_event_t* event) {
+    observer_list_item* cur_observer_item = list;
+    while (cur_observer_item != NULL) {
+        space_observer_notify(cur_observer_item->observer, event);
+        observer_list_item* next_item = cur_observer_item->next;
+    }
+}
+
+typedef struct _atom_list_item {
+    atom_t* atom;
+    struct _atom_list_item* next;
+} atom_list_item;
+
+typedef struct _custom_space_buf {
+    observer_list_item* observers;
+    atom_list_item* atoms;
+    size_t atom_count;
+} custom_space_buf;
+
+void register_observer(void* space_ptr, space_observer_t* observer) {
+    custom_space_buf* space = space_ptr;
+    observer_list_item* new_item = malloc(sizeof(observer_list_item));
+    new_item->observer = observer;
+    new_item->next = NULL;
+    if (space->observers == NULL) {
+        space->observers = new_item;
+    } else {
+        observer_list_item* cur_item = space->observers;
+        while (cur_item->next != NULL) {cur_item = cur_item->next;}
+        cur_item->next = new_item;
+    }
+}
+
+void collect_child_atoms(atom_array_t atoms, void* vec_ptr) {
+    vec_atom_t* vec = vec_ptr;
+    for (int i=0; i<atoms.size; i++) {
+        vec_atom_push(vec, atom_clone(atoms.items[i]));
+    }
+}
+
+void atom_match(const atom_t* q, const atom_t* s, bindings_t* bindings) {
```

Comment:
That's fair.  I wrote this test mainly to make sure I had enough functionality in the C API to implement a space with its own matching behavior - I just didn't have a very creative idea about what behavior to implement.

Reply:
In addition to match_atoms, I realize I also need to export `narrow_vars`.  Which is already public in Rust, so that's reasonable.

And maybe I also need to make a convenience function to extract all the variable atoms from another atom.  (would return itself for variables, nothing for symbols, recursive iteration for expressions, and nothing for grounded atoms)

Reply:
Ok.  I exposed atom_match_atom, atom_iterate & bindings_narrow_vars to the C API, and reworked the test to use them.  This work was committed in e8c269d and 
b4551bb